WalletWasabi.Fluent/ViewModels/Wallets/Home/Tiles/PrivacyRing/PrivacyBarViewModel.cs

```diff
@@ -89,69 +86,95 @@ private void CreateSegments(IExtendedList<PrivacyBarItemViewModel> list, IEnumer
+			rawSegments = rawSegments.Select(x =>
 			{
-				var margin = 2;
-				var amount = coin.Amount.ToDecimal(NBitcoin.MoneyUnit.BTC);
-				var width = Math.Abs((decimal)usableWidth * amount / total);
+				var finalWidth = x.Width < EnlargeThreshold ? x.Width + EnlargeBy : x.Width - reduceBy;
+				return (OwnerPocket: x.OwnerPocket, Coin: x.Coin, Width: finalWidth);
+			}).ToArray();
+		}
 
-				// Artificially enlarge segments smaller than 2 px in order to make them visible.
-				if (width < 2)
-				{
-					width++;
-					margin--;
-				}
+		// Order the coins as they will be shown in the bar.
+		rawSegments = rawSegments
+			.OrderByDescending(x => x.OwnerPocket.Coins.First().HdPubKey.AnonymitySet)
+			.ThenByDescending(x => x.Width)
+			.ToArray();
 
-				yield return new PrivacyBarItemViewModel(this, coin, (double)start, (double)width);
+		var start = 0.0m;
+		foreach (var (_, coin, width) in rawSegments)
+		{
+			yield return new PrivacyBarItemViewModel(this, coin, (double)start, (double)width);
 
-				start += width + margin;
-			}
+			start += width + GapBetweenSegments;
 		}
 	}
 
 	private IEnumerable<PrivacyBarItemViewModel> CreatePocketSegments(IEnumerable<Pocket> pockets)
 	{
-		var total = pockets.Sum(x => Math.Abs(x.Amount.ToDecimal(NBitcoin.MoneyUnit.BTC)));
-		var start = 0.0m;
+		var totalAmount = pockets.Sum(x => Math.Abs(x.Amount.ToDecimal(MoneyUnit.BTC)));
+		var usableWidth = (decimal)Width - (pockets.Count() - 1) * GapBetweenSegments;
 
-		var usableWidth = Width - (pockets.Count() * 2);
+		// Calculate the width of the segments.
+		var rawSegments = pockets.Select(pocket =>
+		{
+			var amount = pocket.Amount.ToDecimal(MoneyUnit.BTC);
+			var width = Math.Abs(usableWidth * amount / totalAmount);
 
-		var usablePockets =
-				pockets.Where(x => x.Coins.Any())
-					   .OrderByDescending(x => x.Coins.First().HdPubKey.AnonymitySet)
-					   .ToList();
+			return (Pocket: pocket, Width: width);
+		}).ToArray();
 
-		foreach (var pocket in usablePockets)
-		{
-			var margin = 2;
-			var amount = pocket.Amount.ToDecimal(NBitcoin.MoneyUnit.BTC);
+		var segmentsToEnlarge = rawSegments.Where(x => x.Width < EnlargeThreshold).ToArray();
 
-			var width = Math.Abs((decimal)usableWidth * amount / total);
+		// Artificially enlarge segments smaller than threshold px in order to make them visible.
+		// Meanwhile decrease those segments that are larger than threshold px in order to fit all in the bar.
+		if (segmentsToEnlarge.Any())
+		{
+			var segmentsToReduce = rawSegments.Except(segmentsToEnlarge);
+			var reduceBy = segmentsToEnlarge.Length * EnlargeBy / segmentsToReduce.Count();
 
-			// Artificially enlarge segments smaller than 2 px in order to make them visible.
-			if (width < 2)
+			rawSegments = rawSegments.Select(x =>
 			{
-				width++;
-				margin--;
-			}
+				var finalWidth = x.Width < EnlargeThreshold ? x.Width + EnlargeBy : x.Width - reduceBy;
+				return (Coin: x.Pocket, Width: finalWidth);
```

Comment:
I am still unure in this part.

- Can the `segmentsToReduce.Count();` be zero?
- What if `reduceBy` is greater than `EnlargeThreshold`? That would cause negative width.

My first impression is that because of the `PrivacyRingMaxItemCount` it is not possible now. But I have to make sure, so will calculate it.

Reply:
https://github.com/zkSNACKs/WalletWasabi/pull/9317/commits/1eaa63ea259b7e7cb9ed0c1fc60bddc764a8f2b9 Fixes this.